src/summaries.rs

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+use abstract_value::{AbstractValue, Path};
+use rpds::{HashTrieMap, List};
+use rustc::hir::def_id::DefId;
+use rustc::ty::TyCtxt;
+use std::collections::HashMap;
+use std::ops::Deref;
+
+/// A summary is a declarative abstract specification of what a function does.
+/// This is calculated once per function and is used by callers of the function.
+/// Callers will specialize this summary by replacing embedded parameter values with the corresponding
+/// argument values and then simplifying the resulting values under the current path condition.
+///
+/// Summaries are stored in a persistent, per project database. When a crate is recompiled,
+/// all summaries arising from the crate are recomputed, but the database is only updated when
+/// a function summary changes. When this happens, all callers of the function need to be reanalyzed
+/// using the new summary, which could result in their summaries being updated, and so on.
+/// In the case of recursive loops, a function summary may need to be recomputed and widened
+/// until a fixed point is reached. Since crate dependencies are acyclic, fixed point computation
```

Comment:
> crate dependencies are acyclic

This is great; it simplifies things so much for us compared to the Z/Infer setting!

Reply:
Indeed. :-)